### ISSUE TITLE
Type Promotion For If-Case Statements

### ISSUE THREAD

**Issue by mcmah309** [dart-lang/language#3505]
For sealed types with two implementations, Dart should be able to promote the type based on if-case statements, like it does with switch statements. Example given
```dart
sealed class Result<S,F> {}

final class Ok<S,F> implements Result<S,F>{
  final S ok;
  
  const Ok(this.ok);
}

final class Err<S,F> implements Result<S,F>{
  final F err;

  const Err(this.err);
}

Result<int,String> okResult() => Ok(1);
```
This works
```dart
Result<int,String> main(){
  Result<int,String> result = okResult();
  switch(result){
    case Err():
      return result;
    case Ok():
  }
  int s = result.ok; // type promotion works
 return Ok(s+1);
}
```
But this does not
```dart
Result<int,String> main(){
  Result<int,String> result = okResult();
  int okVal;
  if(result case Err()){
    return result;
  }
  int s = result.ok; // type promotion does not work here
  return Ok(s + 1);
}
```

**Comment by mcmah309** [comment]
@eernstg based on our current chat here https://github.com/dart-lang/sdk/issues/55104 , I assume the reason this works in the first instance is because specifying all cases of the sealed class triggers exhaustiveness analysis. While this is not currently done for `if case` statements?

**Comment by eernstg** [comment]
Right, if-case chains do not have an exhaustiveness property, no attempt is made to detect any such thing.

I don't think we're going to have it either. @stereotype441, WDYT?

**Comment by lrhn** [comment]
Agree, we do not try to handle exhaustiveness with `if` except for *union types*, where we know that there are precisely two immediate subtypes.

We could do the same for sealed types with two immediate subclasses, but that's a feature which feels partial if it stops working at three subclasses.

We could aim to solve the general problem: Given a sealed type with subtypes `T1`...`Tn`, can we treat it as a union type `T1|...|Tn` and have type checks elimintate some parts of it.
(But then we run into the case of `T4` being another sealed type, and wanting to eliminate that one subclass at a time. Possibly doable, likely not trivial.)
